File: src/main/java/com/affablebean/cart/ShoppingCartItem.java
package com.affablebean.cart;

import java.util.concurrent.atomic.AtomicInteger;

import com.affablebean.domain.Product;

public final class ShoppingCartItem {

	private final Product product;
	private final AtomicInteger quantity = new AtomicInteger(1);

	public ShoppingCartItem(Product product) {
		this.product = product;
	}

	public Product getProduct() {
		return product;
	}

	public short getQuantity() {
		return (short) quantity.get();
	}

	public void setQuantity(short quantity) {
		this.quantity.set(quantity);
	}

	public void incrementQuantity() {
		quantity.incrementAndGet();
	}

	public void decrementQuantity() {
		quantity.decrementAndGet();
	}

	public double getTotal() {
		return quantity.doubleValue() * product.getPrice().doubleValue();
	}

}

File: node/src/session/shoppingCartItem.js
class ShoppingCartItem {
  constructor(product) {
    this.product = product;
    this.quantity = 1;
    this.total = product.price;
  }

  setQuantity(quantity) {
    this.quantity = quantity;
    this.total = quantity * this.product.price;
  }
}

module.exports = {
  ShoppingCartItem
};
